fix: bearer authorization support (#1246)

tkrop opened the pull request:
fixes #1246.

vadeg commented on server/zally-ruleset-zalando/src/test/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRuleTest.kt:
```diff
@@ -48,6 +49,27 @@ class SecureWithOAuth2RuleTest {
+                  type: http
+                  scheme: bearer
+                  bearerFormat: JWT
+                  scopes:
```
This is not available in the API spec.

tkrop replied:
True. Left over from an initial thought how to support this.

tkrop commented on server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRule.kt:
```diff
@@ -14,26 +14,31 @@ import io.swagger.v3.oas.models.security.SecurityScheme
-            .any { SecurityScheme.Type.OAUTH2 == it.type }
+            .orEmpty().any {
+                SecurityScheme.Type.OAUTH2 == it.type ||
+                        (it.type == SecurityScheme.Type.HTTP && it.scheme == "bearer")
```
Seems not to work on my build:

> Task :zally-ruleset-zalando:compileKotlin FAILED
e: /home/krop/workspace/open/zally/server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRule.kt: (37, 59): Unresolved reference: isBearer

vadeg replied:
You need to `import` functions. 
```
import org.zalando.zally.core.util.isBearer
import org.zalando.zally.core.util.isOAuth2
```
Then you can do even like this:
```
            .orEmpty().any {
                it.isOAuth2() || it.isBearer()
            }
```

tkrop commented on server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRule.kt:
```diff
@@ -14,26 +14,31 @@ import io.swagger.v3.oas.models.security.SecurityScheme
-        return if (!oauth2IsUsed) context.violation("API has to be secured by OAuth2", "/components/securitySchemes".toJsonPointer())
+        return if (!oauth2IsUsed) context.violation(
+            "API has to be secured by OAuth2 or Bearer Authentication",
+            "/components/securitySchemes".toJsonPointer()
```
Also fails to compile locally on my build:

> Task :zally-ruleset-zalando:compileKotlin FAILED
e: /home/krop/workspace/open/zally/server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRule.kt: (29, 13): Type mismatch: inferred type is (Mutable)Map<String!, SecurityScheme!>? but Any was expected

tkrop replied:
Does not work, since this creation pointer is actually not related to an existing security schema. I hope, I have made this more clear by renaming the checks accordingly.

tkrop commented on server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureWithOAuth2OrBearerAuthRule.kt:
```diff
@@ -14,26 +14,31 @@ import io.swagger.v3.oas.models.security.SecurityScheme
-            .orEmpty()
+        context.api.components?.securitySchemes?.values?.filter {
+            !(it.type == SecurityScheme.Type.OAUTH2 ||
+                    (it.type == SecurityScheme.Type.HTTP && it.scheme == "bearer"))
```
See above.

tkrop replied:
Works as above mentioned.

tkrop commented:
@vadeg thanks for your review and support.

I have checked all your suggestions and integrated as far as possible. In my latest updated, I adjusted the rule naming to https://github.com/zalando/restful-api-guidelines/pull/666. I also tried to make the function of the checks clearer and added more unit test capture their behavior correctly.

tkrop commented on server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureAllEndpointsRule.kt:
```diff
@@ -12,28 +14,29 @@ import io.swagger.v3.oas.models.security.SecurityScheme
+        }.orEmpty()
 
     @Check(severity = Severity.MUST)
     fun checkUsedScopesAreSpecified(context: Context): List<Violation> {
```
Unfortunately, this check only covers the `clientCredentials` flow. I can imagine, that we also would like to cover the `implicit` and other `OAuth2` flows that are used in Zalando. However, this is not a general problem, since we currently just not check them.

tkrop commented on server/zally-ruleset-zalando/src/main/kotlin/org/zalando/zally/ruleset/zalando/SecureAllEndpointsRule.kt:
```diff
@@ -12,28 +14,29 @@ import io.swagger.v3.oas.models.security.SecurityScheme
 
-        return if (!oauth2IsUsed) context.violation("API has to be secured by OAuth2", "/components/securitySchemes".toJsonPointer())
-        else null
+        return if (valid.isEmpty()) context.violation(
```
It will be rejected as invalid security scheme. It is tested using the `apiKey` security scheme.

jakimarks commented:
👍

zeitlinger commented:
:+1:
